ec/ecresource/deploymentresource/topology_dedicated_masters.go

```diff
@@ -0,0 +1,365 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package deploymentresource
+
+import (
+	"context"
+	"fmt"
+	"github.com/elastic/cloud-sdk-go/pkg/api/deploymentapi/deploymentsize"
+	"github.com/elastic/cloud-sdk-go/pkg/models"
+	es "github.com/elastic/terraform-provider-ec/ec/ecresource/deploymentresource/elasticsearch/v2"
+	"github.com/elastic/terraform-provider-ec/ec/internal/util"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/tfsdk"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-framework/types/basetypes"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+)
+
+var objectAsOptions = basetypes.ObjectAsOptions{UnhandledNullAsEmpty: false, UnhandledUnknownAsEmpty: false}
+
+func UpdateDedicatedMasterTier(
+	ctx context.Context,
+	config tfsdk.Config,
+	plan tfsdk.Plan,
+	privateState PrivateState,
+	resp *resource.ModifyPlanResponse,
+	loadTemplate func() (*models.DeploymentTemplateInfoV2, error),
+) {
+	var esConfig es.ElasticsearchTF
+	resp.Diagnostics.Append(config.GetAttribute(ctx, path.Root("elasticsearch"), &esConfig)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	if !esConfig.MasterTier.IsNull() && !esConfig.MasterTier.IsUnknown() {
+		// Master tier is explicitly configured -> No changes will be made
+		tflog.Debug(ctx, "Skip UpdateDedicatedMasterTier: Master tier has been explicitly configured")
+		return
+	}
+
+	var planElasticsearch es.ElasticsearchTF
+	resp.Diagnostics.Append(plan.GetAttribute(ctx, path.Root("elasticsearch"), &planElasticsearch)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	template, err := loadTemplate()
+	if err != nil {
+		resp.Diagnostics.AddError("Failed to get deployment-template", "Error: "+err.Error())
+		return
+	}
+
+	dedicatedMastersThreshold := getDedicatedMastersThreshold(*template)
+	if dedicatedMastersThreshold == 0 {
+		// No automatic dedicated masters management
+		return
+	}
+
+	nodesInCluster := countNodesInCluster(ctx, planElasticsearch, *template)
+
+	if nodesInCluster < dedicatedMastersThreshold {
+		// Disable master tier
+		if planElasticsearch.MasterTier.IsUnknown() || planElasticsearch.MasterTier.IsNull() {
+			resp.Plan.SetAttribute(ctx,
+				path.Root("elasticsearch").AtName("master"),
+				types.ObjectNull(es.ElasticsearchTopologyAttrs()),
+			)
+		} else {
+			resp.Plan.SetAttribute(ctx,
+				path.Root("elasticsearch").AtName("master").AtName("size"),
+				"0g",
+			)
+		}
```

Comment:
nit-pick: it's a quite long if-else, so it makes sense to return from `if` block and remove `else`. I believe that will be more idiomatic golang.

Reply:
You are right. Another proposal: What if I move the two cases into functions: disable master / enable master. That will make it easier to read